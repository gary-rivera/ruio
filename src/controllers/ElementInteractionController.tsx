const DEFAULT_ROOT_ELEMENT = 'root'

/**
 * Attaches hover and click listeners to elements under the specified root element.
 * When an element is hovered or clicked, the provided callback function is executed.
 *
 * @param {function(HTMLElement): void} onHover - The callback function to be invoked when an element is hovered. Adds styling and applys borders
 * @param {function(HTMLElement): void} onClick - The callback function to be invoked when an element is clicked. Removes styling and borders, cleans up event listeners, and toggles ruioEnabled state to off
 * @returns {function(): void} - A cleanup function that removes all attached event listeners.
 */
export const ElementInteractionController = (
  onHover: (element: HTMLElement) => void,
  onClick: (element: HTMLElement) => void,
) => {
  const rootElement = document.querySelector(`#${DEFAULT_ROOT_ELEMENT}`) as HTMLElement

  if (!rootElement) {
    console.error(
      "[ruio][ElementInteractionController] Root element not found. Make sure your project's root element matches the DEFAULT_ROOT_ELEMENT: ",
      DEFAULT_ROOT_ELEMENT,
    )
    return
  }

  const isValidTarget = (target: HTMLElement) => {
    const isChildOfRoot = target !== rootElement && target.closest(`#${DEFAULT_ROOT_ELEMENT}`)
    return !target.classList.contains('ruio-exclude') && isChildOfRoot && target !== rootElement
    // const isChildOfRoot = target !== rootElement && target.closest(`#${DEFAULT_ROOT_ELEMENT}`)
    // return !target.classList.contains('ruio-exclude') && isChildOfRoot && target !== rootElement
  }

  const originalStyles = new Map<HTMLElement, string>()

  /**
   * Saves the element's current inline styles before applying new hover styles.
   *
   * @param {HTMLElement} target - The element to apply hover styles to.
   */
  const saveOriginalStyles = (target: HTMLElement) => {
    if (!target.classList.contains('ruio-hovered')) {
      originalStyles.set(target, target.getAttribute('style') || '')
      target.classList.add('ruio-hovered')
    }
  }

  /**
   * Restores the original inline styles to the element.
   *
   * @param {HTMLElement} target - The element whose styles should be restored.
   */

  // TODO: we need to ensure the ElemntInteractionController is updated to handle this case
  // requires being more restrictive in the hover and click event listeners, namely our logic behind storing element styles in the originalStyles map section
  const restoreOriginalStyles = (target: HTMLElement) => {
    const originalStyle = originalStyles.get(target)

    originalStyle ? target.setAttribute('style', originalStyle) : target.removeAttribute('style')
    if (originalStyle === '') target.style.backgroundColor = ''

    target.classList.remove('ruio-hovered')
    originalStyles.delete(target)
  }

  const applyHoverStyles = (target: HTMLElement) => {
    if (!originalStyles.has(target)) {
      saveOriginalStyles(target)
    }
    target.style.backgroundColor = 'rgba(153, 181, 214, 0.66)'
  }

  const removeHoverStyles = (target: HTMLElement) => {
    restoreOriginalStyles(target)
  }

  const handleHover = (event: MouseEvent) => {
    const target = event.target as HTMLElement
    if (isValidTarget(target)) {
      applyHoverStyles(target)
      onHover(target)
    }
  }

  const handleMouseOut = (event: MouseEvent) => {
    const target = event.target as HTMLElement
    if (isValidTarget(target) && target.classList.contains('ruio-hovered')) {
      removeHoverStyles(target)
    }
  }

  const handleClick = (event: MouseEvent) => {
    event.preventDefault()

    const target = event.target as HTMLElement
    if (isValidTarget(target)) {
      removeHoverStyles(target)
      onClick(target)
      cleanup()
    }
  }

  document.body.addEventListener('mouseover', handleHover)
  document.body.addEventListener('mouseout', handleMouseOut)
  document.body.addEventListener('click', handleClick)

  const cleanup = () => {
    document.body.removeEventListener('mouseover', handleHover)
    document.body.removeEventListener('mouseout', handleMouseOut)
    document.body.removeEventListener('click', handleClick)
  }

  return cleanup
}
