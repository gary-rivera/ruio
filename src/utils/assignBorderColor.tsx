const colorPalettes: Record<string, string[]> = {
  default: [
    '#ff5832', // Vivid Orange
    '#84a5ad', // Muted Cyan
    '#c17f70', // Soft Brown
    '#5bbfd6', // Bright Cyan
    '#ea6547', // Warm Red-Orange
    '#999999', // Neutral Gray
    '#ad8c84', // Muted Pink
    '#70b2c1', // Light Cyan
    '#d6725b', // Warm Coral
    '#47ccea', // Bright Aqua
  ],
  light: [
    '#FFCCCC', // Light Red
    '#CCE5FF', // Light Blue
    '#CCFFCC', // Light Green
    '#FFF5CC', // Light Yellow
    '#FFCCE5', // Light Pink
    '#CCFFF5', // Light Teal
    '#FFE5CC', // Light Orange
    '#E5CCFF', // Light Purple
    '#FFF0CC', // Light Peach
    '#CCFFEA', // Light Aqua
  ],
  dark: [
    '#990000', // Dark Red
    '#003366', // Dark Blue
    '#006600', // Dark Green
    '#996600', // Dark Yellow/Brown
    '#660033', // Dark Pink
    '#003333', // Dark Teal
    '#994C00', // Dark Orange
    '#330066', // Dark Purple
    '#664400', // Dark Peach/Brown
    '#00664C', // Dark Aqua
  ],
  // didnt realize the negative connotation ._.
  roygbiv: [
    '#FF0000', // Red
    '#FF7F00', // Orange
    '#FFFF00', // Yellow
    '#00FF00', // Green
    '#0000FF', // Blue
    '#4B0082', // Indigo
    '#8B00FF', // Violet
  ],
  neon: [
    '#FF00FF', // Magenta
    '#00FFFF', // Cyan
    '#FFFF00', // Yellow
    '#00FF00', // Bright Green
    '#FF0000', // Bright Red
    '#0000FF', // Bright Blue
    '#FF7F00', // Bright Orange
    // NOTE: pastel pink #FFADAD, pastel peach #FFD6A5, pastel yellow #FDFFB6, pastel green #CAFFBF, pastel blue #9BF6FF, pastel purple #A0C4FF
  ],
  minimalist: [
    '#FFFFFF', // Pure White
    '#000000', // Pure Black
    '#F5F5F5', // White Smoke
    '#1C1C1C', // Very Dark Gray
    '#E5E5E5', // Light Gray
    '#2B2B2B', // Charcoal Black
    '#D3D3D3', // Light Gray (Silver)
    '#4F4F4F', // Dark Gray
    '#FAFAFA', // Very Light White
    '#333333', // Darker Gray
  ],
}

export const getRelativeDepthColor = (theme: keyof typeof colorPalettes, depth: number): string => {
  const palette = colorPalettes[theme]
  if (!palette.length) {
    throw new Error(`The theme '${theme}' does not have any colors defined.`)
  }
  return palette[depth % palette.length]
}
