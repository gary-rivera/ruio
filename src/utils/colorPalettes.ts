export const colorPalettesMap: Record<string, string[]> = {
  default: [
    '#249EA0', // Blue
    '#FAAB36', // Orange
    '#008083', // Cerulean
    '#F78104', // Lime Orange
    '#005F60', // Ruio Cyan
    '#FD5901', // Blood Orange
  ],
  roygbiv: [
    '#CD001A', // Red
    '#EF6A00', // Orange
    '#F2CD00', // Yellow
    '#79C300', // Green
    '#1961AE', // Blue
    '#61007D', // Violet
  ],
  neon: [
    '#FF00FF', // Magenta
    '#00FFFF', // Cyan
    '#FFFF00', // Yellow
    '#00FF00', // Bright Green
    '#FF5F00', // Neon Orange
    '#1E90FF', // Neon Blue
  ],
  minima: [
    '#2F3E46', // Charcoal Gray
    '#CAD2C5', // Pale Green
    '#4A4E69', // Muted Purple
    '#F4F1DE', // Soft Beige
    '#555B6E', // Slate Blue
    '#EAF8EF', // Soft Terracotta
  ],
}

export const getRelativeDepthColor = (colors: string[], currentDepth: number): string => {
  if (!colors.length) {
    throw new Error(`[ruio][utils/colorPalettes] no colors found for the current theme`)
  }
  const index = currentDepth % colors.length
  return colors[index]
}
