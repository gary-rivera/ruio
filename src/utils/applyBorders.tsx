import { getRelativeDepthColor } from '@utils/assignBorderColor'
// only exposed for testing (for now)
export let previouslyAppliedElements: HTMLElement[] = []

export const applyBorders = (element: HTMLElement, depth: number, apply: boolean) => {
  const elements: HTMLElement[] = []

  const traverse = (el: HTMLElement, currentDepth: number) => {
    if (currentDepth > depth) return
    if (el.classList.contains('ruio-exclude') || el.tagName === 'SCRIPT') return

    elements.push(el)

    el.style.outline = apply ? `2px solid ${getRelativeDepthColor('dark', currentDepth)}` : ''

    Array.from(el.children).forEach((child) => {
      if (child instanceof HTMLElement) {
        traverse(child, currentDepth + 1)
      }
    })
  }

  traverse(element, 0)

  if (previouslyAppliedElements.length > 0) {
    previouslyAppliedElements.forEach((el) => {
      // Remove the border styling from elements that are no longer in the list
      if (!elements.includes(el)) el.style.outline = ''
    })
  }

  previouslyAppliedElements = elements
}

// Export a reset function for testing
// tl;dr - encapsulation of logic is important to avoid false positives in tests
export const resetPreviouslyAppliedElements = () => {
  previouslyAppliedElements = []
}
