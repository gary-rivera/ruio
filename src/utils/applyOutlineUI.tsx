import { getRelativeDepthColor } from '@utils/assignBorderColor'

export let previouslyAppliedElements: Set<HTMLElement> = new Set()

// TODO: add root class or id configuration to settings icon modal
// TODO: offer a way to toggle between Sets and Array for previouslyAppliedElements (performance for small vs. large data sets)
// ref. sha for original transition from Array to Set: a1808d5fd72213a86fcc827416e4a6c8891cd1db
export const applyOutlineUI = (element: HTMLElement, depth: number, apply: boolean) => {
  if (element && element.classList.contains('ruio-exclude')) return
  const elements = new Set<HTMLElement>()

  const traverse = (el: HTMLElement, currentDepth: number) => {
    if (!el || currentDepth > depth) return // NOTE: temporary, while we get depth-down figure out

    if (el.classList.contains('ruio-exclude') || el.tagName === 'SCRIPT') return

    elements.add(el)

    // Apply styles only when necessary
    requestAnimationFrame(() => {
      // TODO: not too optimal to constantly lookup rahter than load the hex index based off state
      el.style.outline = apply ? `2px solid ${getRelativeDepthColor('neon', currentDepth)}` : ''
    })

    Array.from(el.children).forEach((child) => {
      if (child instanceof HTMLElement) {
        traverse(child, currentDepth + 1)
      }
    })
  }

  traverse(element, 0)

  requestAnimationFrame(() => {
    previouslyAppliedElements.forEach((el) => {
      // Remove outline if not in the new set of elements
      if (!elements.has(el)) el.style.outline = ''
    })

    previouslyAppliedElements = elements
  })
}

// Export a reset function for testing
export const resetPreviouslyAppliedElements = () => {
  previouslyAppliedElements.clear()
}
