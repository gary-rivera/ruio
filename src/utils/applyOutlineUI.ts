import { getRelativeDepthColor, colorPalettesMap } from '@utils/colorPalettes'

export let previouslyAppliedElements: Set<HTMLElement> = new Set()

// TODO: add root class or id configuration to settings icon modal
// TODO: offer a way to toggle between Sets and Array for previouslyAppliedElements (performance for small vs. large data sets)
// ref. sha for original transition from Array to Set: a1808d5fd72213a86fcc827416e4a6c8891cd1db
export const applyOutlineUI = ({
  element,
  depth,
  currentColorPalette,
  styleProp = 'outline',
}: {
  element: HTMLElement
  depth: number
  currentColorPalette: string
  styleProp?: string
}) => {
  if (!currentColorPalette) {
    console.warn('currentColorPalette is undefined; defaulting to "default" palette.')
    currentColorPalette = 'default'
  }

  const colors = colorPalettesMap[currentColorPalette]
  const elements = new Set<HTMLElement>()

  const traverse = (el: HTMLElement, currentDepth: number) => {
    if (!el || currentDepth > depth) return // NOTE: temporary, while we get depth-down figure out

    if (el.tagName === 'SCRIPT') return

    // NOTE: does this ever get cleaned up? or does the set just grow indefinitely?
    elements.add(el)

    // Apply styles only when necessary
    requestAnimationFrame(() => {
      const outlineColor = getRelativeDepthColor(colors, currentDepth)
      el.style.outline = `2px solid ${outlineColor}`
    })

    Array.from(el.children).forEach((child) => {
      if (child instanceof HTMLElement) {
        traverse(child, currentDepth + 1)
      }
    })
  }

  traverse(element, 0)

  requestAnimationFrame(() => {
    previouslyAppliedElements.forEach((el) => {
      // Remove outline if not in the new set of elements
      if (!elements.has(el)) el.style.outline = ''
    })

    previouslyAppliedElements = elements
  })
}

// Export a reset function for testing
export const resetPreviouslyAppliedElements = () => {
  previouslyAppliedElements.clear()
}
