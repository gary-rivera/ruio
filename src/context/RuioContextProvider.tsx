import React, {
  createContext,
  useState,
  useEffect,
  ReactNode,
  useContext,
  useCallback,
  useMemo,
  useRef,
} from 'react'
import { applyBorders } from '../utils/applyBorders'
import { ElementInteractionController } from '../controllers/ElementInteractionController'
import ControlPanel from '@components/ControlPanel'

interface RuioContextProps {
  ruioEnabled: boolean // is ruio related state/interactions enabled
  setRuioEnabled: React.Dispatch<React.SetStateAction<boolean>> // toggle ruio related state/interactions
  depth: number // depth of the amount of elements to apply borders to
  setDepth: React.Dispatch<React.SetStateAction<number>> // set the depth of the amount of elements to apply borders to
  selectedRootElement: HTMLElement | null // the root element that is selected (defaults to where div.body#root)
  isElementSelectionModeActive: boolean // is element selection mode active -- aka are there hover and click events drilled into the DOM
  setIsElementSelectionModeActive: React.Dispatch<React.SetStateAction<boolean>> // toggle element selection mode
  toggleElementSelectionMode: () => void
}

const RuioContext = createContext<RuioContextProps | undefined>(undefined)

export const RuioContextProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [ruioEnabled, setRuioEnabled] = useState(false)
  const [depth, setDepth] = useState(1)
  const [selectedRootElement, setSelectedRootElement] = useState<HTMLElement | null>(null)
  const [isElementSelectionModeActive, setIsElementSelectionModeActive] = useState(false)

  const controlPanelRef = useRef<HTMLDivElement | null>(null)

  useEffect(() => {
    if (ruioEnabled && controlPanelRef.current) {
      controlPanelRef.current.style.display = 'block'
    } else if (controlPanelRef.current) {
      controlPanelRef.current.style.display = 'none'
    }
  }, [ruioEnabled])

  // TODO: use this to store the previous selected root element for when a user exits element selection mode without picking an element
  const previousSelectedRootElementRef = useRef<HTMLElement | null>(null)
  // const previousSelectedRootElementRef = useRef<HTMLElement | null>(null)

  /**
   * Triggers element selection mode by toggling the active state.
   * Wrapped in useCallback to maintain referential equality in contextValue.
   */
  const toggleElementSelectionMode = useCallback(() => {
    setIsElementSelectionModeActive((prev) => !prev)
  }, [])

  useEffect(() => {
    if (isElementSelectionModeActive) {
      // trigger ElementInteractionController (add click and hover events to DOM used for root element selection)
      const cleanupElementSelectionEvents = ElementInteractionController(
        (element: HTMLElement) => {
          applyBorders(element, depth, ruioEnabled)
        },
        (element: HTMLElement) => {
          setSelectedRootElement(element)
          setIsElementSelectionModeActive(false)
        },
      )

      return () => {
        if (cleanupElementSelectionEvents) {
          cleanupElementSelectionEvents()
        }
      }
    }
    // NOTE: should the cleanup function be saved to a ref to be called later when element selection mode is exited?
    // when an element is selected: yes [] no []
    // when an element is not selected: yes [] no []
  }, [isElementSelectionModeActive, depth, ruioEnabled])

  useEffect(() => {
    if (selectedRootElement) {
      applyBorders(selectedRootElement, depth, ruioEnabled)
    }
  }, [depth, selectedRootElement]) // NOTE: did include ruioEnabled at one point -- might need to add back

  const contextValue = useMemo(
    () => ({
      ruioEnabled,
      setRuioEnabled,
      depth,
      setDepth,
      selectedRootElement,
      isElementSelectionModeActive,
      setIsElementSelectionModeActive,
      toggleElementSelectionMode,
    }),
    [ruioEnabled, depth, selectedRootElement, toggleElementSelectionMode],
  )

  return (
    <RuioContext.Provider value={contextValue}>
      <ControlPanel ref={controlPanelRef} />
      {children}
    </RuioContext.Provider>
  )
}

/**
 * Custom hook to access the Ruio context.
 */
export const useRuioContext = () => {
  const context = useContext(RuioContext)
  if (!context) {
    throw new Error('[RuioContextProvider] useRuio must be used within RuioProvider')
  }
  return context
}
