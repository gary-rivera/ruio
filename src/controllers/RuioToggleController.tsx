import { useRuioContext } from '@root/context/RuioContextProvider'

import buttonStyles from '../styles/Button.module.css'
import divStyles from '../styles/Div.module.css'
import svgStyles from '../styles/SVG.module.css'

function RuioToggleController() {
  const { ruioEnabled, setRuioEnabled } = useRuioContext()

  return (
    <div className={`ruio-exclude ${divStyles['ruio-toggle-container']}`}>
      <button
        className={`
        ruio-exclude
        ${buttonStyles['ruio-btn']}
        ${buttonStyles['ruio-btn-primary']} ${ruioEnabled ? buttonStyles['ruio-logo-btn-enabled'] : buttonStyles['ruio-logo-btn-disabled']}`}
        onClick={() => {
          setRuioEnabled(!ruioEnabled)
        }}
      >
        <div
          className={`${divStyles['ruio-logo-div']} ${ruioEnabled ? divStyles['ruio-logo-div-active'] : divStyles['ruio-logo-div-inactive']}`}
        >
          <svg
            className={`ruio-exclude ${svgStyles['ruio-logo-svg']}`}
            viewBox="0 0 208 176"
            fill="none"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              fillRule="evenodd"
              clipRule="evenodd"
              d="M173.242 19.5556H35.2418C25.7149 19.5556 17.9918 28.3109 17.9918 39.1111V136.889C17.9918 147.689 25.7149 156.444 35.2418 156.444H59.4968V56.8272L63.5747 53.959C76.3361 44.9831 109.235 32.3035 140.023 53.959L140.382 54.2116L140.718 54.5029C145.937 59.0309 152.719 67.5177 154.79 78.5538C155.868 84.2978 155.634 90.6403 153.274 97.053C150.981 103.28 146.895 109.009 141.036 114.168C138.393 116.83 134.291 119.017 129.416 120.468C127.801 120.949 126.037 121.369 124.122 121.704L148.87 156.444H173.242C182.769 156.444 190.492 147.689 190.492 136.889V39.1111C190.492 28.3109 182.769 19.5556 173.242 19.5556ZM91.8352 101.581L128.236 156.444H128.247L139.537 176H173.242C192.296 176 207.742 158.489 207.742 136.889V39.1111C207.742 17.5106 192.296 0 173.242 0H35.2418C16.188 0 0.741821 17.5106 0.741821 39.1111V136.889C0.741821 158.489 16.188 176 35.2418 176H59.4968H76.7468V156.444V68.0911C94.0474 58.8496 133.144 57.18 137.918 82.6223C138.333 84.8357 138.246 87.0982 137.354 89.522C135.26 95.2123 130.644 99.8841 125.035 101.554C121.597 102.577 116.36 103.378 109.091 102.845L91.8352 101.581Z"
            />
          </svg>
        </div>
      </button>
      <div
        className={`${divStyles['ruio-logo-div-bg']} ${ruioEnabled ? divStyles['ruio-logo-div-bg-active'] : divStyles['ruio-logo-div-bg-inactive']}`}
      />
    </div>
  )
}

export default RuioToggleController
