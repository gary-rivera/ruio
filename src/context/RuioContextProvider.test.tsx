import { render, screen, act } from '@testing-library/react'
import { RuioContextProvider, useRuioContext } from './RuioContextProvider'
import { applyBorders } from '../utils/applyBorders'
import { ElementInteractionController } from '../controllers/ElementInteractionController'
import userEvent from '@testing-library/user-event'
import { waitFor } from '@testing-library/react'

// mocks source
jest.mock('../utils/applyBorders')
jest.mock('../controllers/ElementInteractionController')

// mocks target
const mockedElementInteractionController = ElementInteractionController as jest.MockedFunction<
  typeof ElementInteractionController
>
const mockedApplyBorders = applyBorders as jest.MockedFunction<typeof applyBorders>

const TestComponent = () => {
  const {
    ruioEnabled,
    setRuioEnabled,
    depth,
    setDepth,
    isElementSelectionModeActive,
    setIsElementSelectionModeActive,
    selectedRootElement,
  } = useRuioContext()

  return (
    <div>
      <div data-testid="ruioEnabled">{ruioEnabled ? 'Enabled' : 'Disabled'}</div>
      <div data-testid="depth">{depth}</div>
      <div data-testid="selectedRootElement">{selectedRootElement?.tagName || 'None'}</div>
      <button
        onClick={() => {
          setIsElementSelectionModeActive(!isElementSelectionModeActive)
        }}
      >
        Select Element Mode
      </button>
      <button onClick={() => setRuioEnabled(true)}>Enable Borders</button>
      <button onClick={() => setDepth(5)}>Set Depth to 5</button>
    </div>
  )
}

describe('RuioContextProvider', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  test('matches the snapshot for initial render', () => {
    const { container } = render(
      <RuioContextProvider>
        <TestComponent />
      </RuioContextProvider>,
    )
    expect(container).toMatchSnapshot()
  })

  test('renders the provider without crashing', () => {
    render(
      <RuioContextProvider>
        <div>Test</div>
      </RuioContextProvider>,
    )
    expect(screen.getByText('Test')).toBeInTheDocument()
  })

  test('should trigger element selection mode and call ElementInteractionController', async () => {
    const cleanupMock = jest.fn()
    const mockElement = document.createElement('div')

    mockedElementInteractionController.mockImplementation((callback) => {
      return cleanupMock
    })

    act(() => {
      render(
        <RuioContextProvider>
          <TestComponent />
        </RuioContextProvider>,
      )
    })

    const selectElementButton = screen.getByText('Select Element Mode')

    await act(async () => {
      await userEvent.click(selectElementButton)
    })

    expect(mockedElementInteractionController).toHaveBeenCalledTimes(1)

    await act(async () => {
      const callback = mockedElementInteractionController.mock.calls[0][0]
      callback(mockElement)
    })

    await waitFor(() => {
      expect(screen.getByTestId('selectedRootElement').textContent).toBe(mockElement.tagName)
    })
  })

  test('should update ruioEnabled state when triggered', async () => {
    render(
      <RuioContextProvider>
        <TestComponent />
      </RuioContextProvider>,
    )

    const enableBordersButton = screen.getByText('Enable Borders')

    await act(async () => {
      await userEvent.click(enableBordersButton)
    })

    expect(screen.getByTestId('ruioEnabled').textContent).toBe('Enabled')
  })

  test('should update depth state when triggered', async () => {
    render(
      <RuioContextProvider>
        <TestComponent />
      </RuioContextProvider>,
    )

    const setDepthButton = screen.getByText('Set Depth to 5')

    await act(async () => {
      await userEvent.click(setDepthButton)
    })

    expect(screen.getByTestId('depth').textContent).toBe('5')
  })

  test('should call cleanup function when selection mode is triggered', async () => {
    const cleanupMock = jest.fn()

    mockedElementInteractionController.mockReturnValue(cleanupMock)

    render(
      <RuioContextProvider>
        <TestComponent />
      </RuioContextProvider>,
    )

    const selectElementButton = screen.getByText('Select Element Mode')

    await act(async () => {
      await userEvent.click(selectElementButton)
    })

    expect(cleanupMock).not.toHaveBeenCalled()

    cleanupMock()

    expect(cleanupMock).toHaveBeenCalled()
  })

  test('should display "None" if no element is selected', () => {
    render(
      <RuioContextProvider>
        <TestComponent />
      </RuioContextProvider>,
    )

    expect(screen.getByTestId('selectedRootElement').textContent).toBe('None')
  })

  test('should call applyBorders with correct arguments when borders are enabled', async () => {
    const mockElement = document.createElement('div')
    mockedElementInteractionController.mockImplementation((callback: (element: HTMLElement) => void) => {
      callback(mockElement)
      return jest.fn()
    })

    render(
      <RuioContextProvider>
        <TestComponent />
      </RuioContextProvider>,
    )

    const enableBordersButton = screen.getByText('Enable Borders')

    await act(async () => {
      await userEvent.click(enableBordersButton)
    })

    const selectElementButton = screen.getByText('Select Element Mode')

    await act(async () => {
      await userEvent.click(selectElementButton)
    })

    expect(mockedApplyBorders).toHaveBeenCalledWith(mockElement, 1, true)
  })

  // NOTE: consider enabling this test for possible regression checks (not working rn)
  // test('should call cleanup function when component unmounts', async () => {
  //   const cleanupMock = jest.fn()

  //   mockedElementInteractionController.mockImplementation(() => {
  //     console.log('ElementInteractionController called, returning cleanupMock')
  //     return cleanupMock
  //   })

  //   const { unmount } = render(
  //     <RuioContextProvider>
  //       <TestComponent />
  //     </RuioContextProvider>,
  //   )

  //   const selectElementButton = screen.getByText('Select Element Mode')

  //   await act(async () => {
  //     userEvent.click(selectElementButton)
  //   })

  //   await act(async () => {
  //     unmount()
  //   })

  //   console.log('Unmounting the component, checking cleanupMock')

  //   expect(cleanupMock).toHaveBeenCalledTimes(1)
  // })

  test('should throw error if useRuioContext is used outside provider', () => {
    const consoleErrorMock = jest.spyOn(console, 'error').mockImplementation(() => {})

    const TestInvalidComponent = () => {
      const { ruioEnabled } = useRuioContext()
      return <div>{ruioEnabled ? 'Enabled' : 'Disabled'}</div>
    }

    expect(() => render(<TestInvalidComponent />)).toThrow(
      '[RuioContextProvider] useRuio must be used within RuioProvider',
    )

    consoleErrorMock.mockRestore()
  })
})
