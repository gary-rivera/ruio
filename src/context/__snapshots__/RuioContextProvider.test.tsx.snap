// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RuioContextProvider matches the snapshot for initial render 1`] = `
<div>
  <div
    class="ruio-exclude"
    data-testid="ruio-control-panel"
    style="position: fixed; bottom: 80px; right: 20px; z-index: 9999; background-color: white; padding: 10px; border: 1px solid black; display: none;"
  >
    <label
      class="ruio-exclude"
    >
      Depth:
      <input
        class="ruio-exclude"
        max="10"
        min="1"
        type="number"
        value="1"
      />
    </label>
    <div
      class="ruio-exclude"
    >
      <button
        class="ruio-exclude"
        disabled=""
      >
        Select Element
      </button>
      <button>
        Enable Borders
      </button>
    </div>
  </div>
  <div>
    <div
      data-testid="ruioEnabled"
    >
      Disabled
    </div>
    <div
      data-testid="depth"
    >
      1
    </div>
    <div
      data-testid="selectedRootElement"
    >
      None
    </div>
    <button
      data-testid="select-element-mode"
    >
      Select Element Mode
    </button>
    <button
      data-testid="enable-borders"
    >
      Enable Borders
    </button>
    <button
      data-testid="set-depth"
    >
      Set Depth to 5
    </button>
  </div>
</div>
`;
